### PR TITLE
Add semeval datasets. Fix #18

### LINKED ISSUE #18: SemEval 2016/2017 Task 3, English Subtask A unannotated datasets and English Subtask B datasets
## Introduction
I converted SemEval 2016 and 2017 question answering datasets into JSON for ease of use. The original datasets are in XML and scattered across several ZIP archives. The JSON files are going to be immediately used in the Gensim documentation for the Soft Cosine Measure (see the respective [pull request](https://github.com/RaRe-Technologies/gensim/pull/1827)).

## Datasets
- [semeval-2016_2017-task3-subtaskA-unannotated-english.json.gz](https://github.com/Witiko/semeval-2016_2017-task3-subtaskA-unannotated-english/releases/download/2018-01-29/semeval-2016_2017-task3-subtaskA-unannotated-english.json.gz) (231M) – Example:
```json
    [
        {

            "THREAD_SEQUENCE": "Q1", 
            "RelQuestion": {
                "RELQ_CATEGORY": "Politics", 
                "RELQ_DATE": "2009-12-14 11:58:33", 
                "RELQ_ID": "Q1", 
                "RELQ_USERID": "U2", 
                "RELQ_USERNAME": "anonymous", 
                "RelQBody": "The state of Internet in Thailand:IT Minsitry blocks CNN; Facebook; Yahoo; Flickr Thai Immigration website listed as dangerousFull story: http://www.thaivisa.com/forum/Thai-Govt-Blocks-Cnn-Yahoo-Financ-t321851.html", 
                "RelQSubject": "Thailand:IT Minsitry blocks CNN; Facebook;"
            }, 
            "RelComments": [
                {
                    "RELC_DATE": "2009-12-14 12:00:59", 
                    "RELC_ID": "Q1_C1", 
                    "RELC_USERID": "U210", 
                    "RELC_USERNAME": "DaRuDe", 
                    "RelCText": "have they blocked porn??? <img src=\"http://www.qatarliving.com/files/images/Da.gif\">"
                }, 
                {
                    "RELC_DATE": "2009-12-14 12:07:04", 
                    "RELC_ID": "Q1_C2", 
                    "RELC_USERID": "U2", 
                    "RELC_USERNAME": "anonymous", 
                    "RelCText": "like trying to contain a tsunami with a hand towel ************************************ I'm Jack's complete lack of surprise"
                }, 
                {
                    "RELC_DATE": "2009-12-14 12:09:23", 
                    "RELC_ID": "Q1_C3", 
                    "RELC_USERID": "U114", 
                    "RELC_USERNAME": "GodFather.", 
                    "RelCText": "oops double post.. ----------------- \"HE WHO DARES WINS\" Derek Edward Trotter"
                }, 
```
- [semeval-2016_2017-task3-subtaskB-english.json.gz](https://github.com/Witiko/semeval-2016_2017-task3-subtaskB-english/releases/download/2018-01-29/semeval-2016_2017-task3-subtaskB-english.json.gz) (6.05M) – Example:
    ```json
    {
        "2016-dev": [
            {
                "ORGQ_ID": "Q268", 
                "OrgQBody": "Which is a good bank as per your experience in Doha", 
                "OrgQSubject": "Good Bank", 
                "Threads": [
                    { 
                        "THREAD_SEQUENCE": "Q268_R4",
                        "RelQuestion": {
                            "RELQ_CATEGORY": "Advice and Help", 
                            "RELQ_DATE": "2013-05-02 19:43:00", 
                            "RELQ_ID": "Q268_R4", 
                            "RELQ_RANKING_ORDER": "4", 
                            "RELQ_RELEVANCE2ORGQ": "PerfectMatch", 
                            "RELQ_USERID": "U4882", 
                            "RELQ_USERNAME": "ankukuma", 
                            "RelQBody": "Hi Guys; I need to open a new bank accoount. Which is the best bank in Qatar ? I assume all of them will roughly be the same; but stll which has a slight edge (Money transfer; benifits etc) Thanks !!!", 
                            "RelQSubject": "Best Bank"
                        }, 
                        "RelComments": [
                            {
                                "RELC_DATE": "2013-05-03 07:23:20", 
                                "RELC_ID": "Q268_R4_C1", 
                                "RELC_RELEVANCE2ORGQ": "Good", 
                                "RELC_RELEVANCE2RELQ": "Good", 
                                "RELC_USERID": "U594", 
                                "RELC_USERNAME": "Dilgeer", 
                                "RelCText": "Commercial bank/IBQ"
                            }, 
                            {
                                "RELC_DATE": "2013-05-03 12:58:13", 
                                "RELC_ID": "Q268_R4_C2", 
                                "RELC_RELEVANCE2ORGQ": "Good", 
                                "RELC_RELEVANCE2RELQ": "Good", 
                                "RELC_USERID": "U979", 
                                "RELC_USERNAME": "Speedysid", 
                                "RelCText": "The best bank in Qatar for you would be the one that fits in your requirements.I suggest you visit the major banks here; and approach the Customer Relations person there to guide you with the facilities the bank offers. They include: -Current Accounts facilities -Savings Account facilities - Money Transfer (However; I highly recommend using the bank transfer only in emergency cases. There are money transfer agents which offer better exchange rates; and lower service fees) - Tie-ups with any bank in your home country to ease transfers"
                            }, 
    ```

## Papers
- [SemEval-2016 Task 3: Community Question Answering](http://alt.qcri.org/semeval2016/task3/data/uploads/semeval2016-task3-report.pdf) ([website](http://alt.qcri.org/semeval2016/task3/))
- [SemEval-2017 Task 3: Community Question Answering](http://alt.qcri.org/semeval2017/task3/data/uploads/semeval2017-task3.pdf) ([website](http://alt.qcri.org/semeval2017/task3/))

## Code
- [witiko/semeval-2016_2017-task3-subtaskA-unannotated-english](https://github.com/Witiko/semeval-2016_2017-task3-subtaskA-unannotated-english)
- [witiko/semeval-2016_2017-task3-subtaskB-english](https://github.com/Witiko/semeval-2016_2017-task3-subtaskB-english)

## License
These are the licensing notices found in the individual ZIP files with the original XML datasets:
- [semeval2016-task3-cqa-ql-traindev-v3.2.zip](http://alt.qcri.org/semeval2016/task3/data/uploads/semeval2016-task3-cqa-ql-traindev-v3.2.zip)
    > These datasets are free for general research use.
- [semeval2017_task3_test.zip](http://alt.qcri.org/semeval2017/task3/data/uploads/semeval2017_task3_test.zip)
    > - the scripts and all files released for the task are free for general research use 
    > - you should use the following citation in your publications whenever using these resources:
    > 
    >       @InProceedings{SemEval-2017:task3,
    >          author    = {Nakov, Preslav and Hoogeveen, Doris and M\`{a}rquez, Llu\'{i}s and Moschitti, Alessandro and Mubarak, Hamdy and Baldwin, Timothy and Verspoor, Karin},
    >          title     = {{SemEval}-2017 Task 3: Community Question Answering},
    >          booktitle = {Proceedings of the 11th International Workshop on Semantic Evaluation},
    >          series    = {SemEval '17},
    >          month     = {August},
    >          year      = {2017},
    >          address   = {Vancouver, Canada},
    >          publisher = {Association for Computational Linguistics},
    >        }

### PR DESCRIPTION
**Pull request by menshikh-iv**

### CODE REVIEW

File: list.json [hunk 1]
```diff
@@ -1,5 +1,51 @@
...
+			"file_name": "semeval-2016-2017-task3-subtaskB-eng.gz",
+			"read_more": ["http://alt.qcri.org/semeval2017/task3/", "http://alt.qcri.org/semeval2017/task3/data/uploads/semeval2017-task3.pdf", "https://github.com/RaRe-Technologies/gensim-data/issues/18", "https://github.com/Witiko/semeval-2016_2017-task3-subtaskB-english"],
+			"parts": 1
+		},
+		"semeval-2016-2017-task3-subtaskA-unannotated-eng": {
+			"num_records": 189941,
+			"record_format": "dict",
+			"file_size": 234373151,
+			"reader_code": "https://github.com/RaRe-Technologies/gensim-data/releases/download/semeval-2016-2017-task3-subtaskA-unannotated-eng/__init__.py",
+			"license": "These datasets are free for general research use.",
```
Comment (piskvorky) [inline list.json]: Do we have a link supporting this conclusion?

**Comment by menshikh-iv** [comment]
Last needed changes

- [ ] "description" for both datasets
- [ ] "fields" for `taskB`
- [ ] Text for release notes
  - [ ] Table with main metrics
  - [ ] Full code example (especcially for `taskB` with evaluation, we can combine it with `taskA`
- [ ] re-generate table for `README`

CC: @Witiko

File: list.json (hunk 1 above)
Comment (Witiko) [reply list.json]: The datasets were put together from the following files:
1. [QL-unannotated-data-subtaskA.xml.zip](http://alt.qcri.org/semeval2016/task3/data/uploads/QL-unannotated-data-subtaskA.xml.zip)
2. [semeval2016-task3-cqa-ql-traindev-v3.2.zip](http://alt.qcri.org/semeval2016/task3/data/uploads/semeval2016-task3-cqa-ql-traindev-v3.2.zip)
3. [semeval2016_task3_test.zip](http://alt.qcri.org/semeval2016/task3/data/uploads/semeval2016_task3_test.zip)
4. [semeval2017_task3_test.zip](http://alt.qcri.org/semeval2017/task3/data/uploads/semeval2017_task3_test.zip)

Files 2 and 4 contain explicit license notices – see the “License” section of #18. Files 1 and 3 contain no licensing notices, so technically all rights are reserved. However, to me it seems like an oversight on the side of the task authors who did not include instructions to each of the several ZIP archives.

**Comment by Witiko** [comment]
> Table with main metrics

According to Section 5 of [the 2016 task paper](http://alt.qcri.org/semeval2016/task3/data/uploads/semeval2016-task3-report.pdf) linked in section “Papers” of #18, the main evaluation metric is MAP (Mean Average Precision). Supplementary measures include Mean Reciprocal Rank (MRR), Average Recall (AvgRec), Precision, Recall, F1, and Accuracy.

**Comment by Witiko** [comment]
Along with the updated datatype of the `RELQ_RANKING_ORDER` field, which I proposed in #18, I have the following name change to propose:

- drop the `-eng` suffix; despite my original belief, the rest of the name should be sufficient to identify the language of the dataset,
- change the `taskB` suffix to `taskBC`; it appears that the dataset can also be used for subtask C.

I apologize for these changes.

**Comment by Witiko** [comment]
> "description" for both datasets

SemEval 2016 / 2017 Task 3 Subtask A unannotated dataset contains 189,941 questions and 1,894,456 comments in English collected from the Community Question Answering (CQA) web forum of [Qatar Living](http://www.qatarliving.com/forum). These can be used as a corpus for language modelling.

SemEval 2016 / 2017 Task 3 Subtask B and C datasets contain train+development (317 original questions, 3,169 related questions, and 31,690 comments), and test datasets in English. The description of the tasks and the collected data is given in sections 3 and 4.1 of the 2016 task paper linked in section “Papers” of #18.

**Comment by Witiko** [comment]
> "fields" for taskB

The main data field for Subtask B is `RELQ_RELEVANCE2ORGQ`, and the main data filed for Subtask C is `RELC_RELEVANCE2ORGQ`. The purpose of the numerous supplementary fields is described in Section 4.1 of the [2016 task paper](http://alt.qcri.org/semeval2016/task3/data/uploads/semeval2016-task3-report.pdf) linked in section “Papers” of #18.

**Comment by Witiko** [comment]
> Full code example

For Subtask A, building a corpus serves as a good code example:

```python
import gensim.downloader as api
from gensim.utils import simple_preprocess

corpus = []
for thread in api.load("semeval-2016-2017-task3-subtaskA-unannotated"):
    corpus.append(simple_preprocess(thread["RelQuestion"]["RelQSubject"]))
    corpus.append(simple_preprocess(thread["RelQuestion"]["RelQBody"]))
    for relcomment in thread["RelComments"]:
        corpus.append(simple_preprocess(relcomment["RelCText"]))
```

The below code example for Subtask B takes the corpus we have just built and uses it to construct a dictionary. For each original thread, we then extract the question from the original thread and compare it against the questions in the related threads, producing a ranking. We evaluate the quality of our rankings using the Mean Average Precision (MAP) evaluation metric.

```python
import gensim.downloader as api
from gensim.corpora import Dictionary
from gensim.similarities import MatrixSimilarity
from gensim.utils import simple_preprocess
import numpy as np

dictionary = Dictionary(corpus)
datasets = api.load("semeval-2016-2017-task3-subtaskBC")

def produce_test_data(dataset):
    for orgquestion in datasets[dataset]:
        query = dictionary.doc2bow(
            simple_preprocess(orgquestion["OrgQSubject"]) \
            + simple_preprocess(orgquestion["OrgQBody"]))
        documents = [
            dictionary.doc2bow(
                simple_preprocess(thread["RelQuestion"]["RelQSubject"]) \
                + simple_preprocess(thread["RelQuestion"]["RelQBody"]))
            for thread in orgquestion["Threads"]]
        relevance = [
            thread["RelQuestion"]["RELQ_RELEVANCE2ORGQ"] \
                in ("PerfectMatch", "Relevant")
            for thread in orgquestion["Threads"]]
        yield query, documents, relevance

def evaluate(dataset):
    results = []
    for query, documents, relevance in produce_test_data(dataset):
        index = MatrixSimilarity(documents, num_features=len(dictionary))
        similarities = index[query]
        assert len(similarities) == len(documents)
        precision = [
            (num_correct + 1) / (num_total + 1) \
            for num_correct, num_total in enumerate(
                num_total for num_total, (_, relevant) in enumerate(
                    sorted(zip(similarities, relevance), reverse=True))\
                    if relevant)]
        average_precision = np.mean(precision) if precision else 0.0
        results.append(average_precision)
    return np.mean(results) * 100.0

for dataset in ("2016-dev", "2016-test", "2017-test"):
    print("MAP score on the %s dataset:\t%.02f" % (dataset, evaluate(dataset)))
```

The above code produces the following output for me:
```
MAP score on the 2016-dev dataset:  66.87
MAP score on the 2016-test dataset: 69.51
MAP score on the 2017-test dataset: 41.06
```

**Comment by Witiko** [comment]
> re-generate table for `README`

Can I help with this?

File: README.md [hunk 2]
```diff
@@ -123,10 +127,9 @@ To load a model or corpus, use either the Python or command line interface of [G
...
-
+| word2vec-ruscorpora-300 | 184973 | 198 MB | Russian National Corpus (about 250M words) | <ul><li>https://www.academia.edu/24306935/WebVectors_a_Toolkit_for_Building_Web_Interfaces_for_Vector_Semantic_Models</li> <li>http://rusvectores.org/en/</li> <li>https://github.com/RaRe-Technologies/gensim-data/issues/3</li></ul> | Word2vec Continuous Skipgram vectors trained on full Russian National Corpus (about 250M words). The model contains 185K words. | <ul><li>dimension - 300</li> <li>window_size - 10</li></ul> | The corpus was lemmatized and tagged with Universal PoS | https://creativecommons.org/licenses/by/4.0/deed.en |
 
-(this table is generated automatically by [generate_table.py](https://github.com/RaRe-Technologies/gensim-data/blob/master/generate_table.py) based on [list.json](https://github.com/RaRe-Technologies/gensim-data/blob/master/list.json))
+(generated by generate_table.py based on list.json)
```
Comment (piskvorky) [inline README.md]: Why this change? It's better with the links. More concrete.

Comment (menshikh-iv) [reply README.md]: This is part of auto-generated stuff, In general, the script has no idea where exactly it is contained on the GitHub (these links were spelled out by hand).

Comment (piskvorky) [reply README.md]: Aha. Can you add it back? (to the script)

Comment (menshikh-iv) [reply README.md]: @piskvorky links never were part of the script, I can add it only manually.

Comment (menshikh-iv) [reply README.md]: UPD: Done 80ad749dd609f407f9aedd9ab1328924a18b95e3

Comment (piskvorky) [reply README.md]: Not a good idea to do it manually. Please add to the script.

Comment (menshikh-iv) [reply README.md]: @piskvorky please look again to https://github.com/RaRe-Technologies/gensim-data/pull/19#discussion_r166830555

Comment (piskvorky) [reply README.md]: What do you mean? Add the links to the scripts, so you don't have to do this manually the next time(s).

Comment (menshikh-iv) [reply README.md]: @piskvorky imagine: I moved/renamed/removed this script or lists.json, the link will be broken.

Done: https://github.com/RaRe-Technologies/gensim-data/commit/fcc89c2d6832cbf19fdc1b896451dc3432631833

Comment (piskvorky) [reply README.md]: Thanks.
